### PR TITLE
`TextRenderer` support custom font asset

### PR DESCRIPTION
**Pull request by singlecoder**
### Please check if the PR fulfills these requirements

- [ ] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: packages/core/src/2d/text/Font.ts [hunk 1]
```diff
@@ -14,9 +15,26 @@ export class Font extends RefObject {
    * Create a font from OS.
    * @param engine - Engine to which the font belongs
    * @param name - The name of font
+   * @param fontUrl - The font url to register, if not, will use system font
    * @returns The font object has been create
    */
-  static createFromOS(engine: Engine, name: string = ""): Font {
+  static async create(engine: Engine, name: string, fontUrl: string = ""): Promise<Font> {
```
Comment (gz65555) [inline packages/core/src/2d/text/Font.ts:21]: Use function overloading

File: packages/core/src/2d/text/TextUtils.ts [hunk 2]
```diff
@@ -248,6 +248,69 @@ export class TextUtils {
...
+        const fontDesc = `40px ${fontName}`;
+        context.font = fontDesc;
+        const textWidth = context.measureText(testString).width;
+
+        const fontStyle = document.createElement("style");
+        fontStyle.type = "text/css";
+        const { body } = document;
+        body.appendChild(fontStyle);
+        fontStyle.textContent = `@font-face { font-family:${fontName}; src:url("${fontUrl}");}`;
+        let div = document.createElement("div");
```
Comment (gz65555) [inline packages/core/src/2d/text/TextUtils.ts:280]: use `const `

Comment (singlecoder) [reply packages/core/src/2d/text/TextUtils.ts:280]:
```typescript
if (div && div.parentNode) {
  div.parentNode.removeChild(div);
  div = null;
}
```
div will be set to null

File: packages/core/src/2d/text/Font.ts (hunk 1 above)
Comment (singlecoder) [reply packages/core/src/2d/text/Font.ts:21]: Just one api for user, create, createFromOS should be internal

File: packages/core/src/2d/text/TextUtils.ts [hunk 3]
```diff
@@ -247,68 +247,12 @@ export class TextUtils {
...
   static async registerFont(fontName: string, fontUrl: string): Promise<boolean> {
-    if (FontFace) {
-      return this._registerFontWithFontFace(fontName, fontUrl);
-    } else {
-      return this._registerFontWithCSS(fontName, fontUrl);
-    }
-  }
-
-  private static async _registerFontWithFontFace(fontName: string, fontUrl: string): Promise<boolean> {
     const fontFace = new FontFace(fontName, `url(${fontUrl})`);
```
Comment (eyworldwide) [inline packages/core/src/2d/text/TextUtils.ts:250]: if `FontFace` API not exists, you should use default system font instead. The downgrade logic should be added.

File: packages/core/src/2d/text/Font.ts [hunk 4]
```diff
@@ -11,25 +11,44 @@ export class Font extends RefObject {
...
-   * Create a font from OS.
+   * Create a font.
    * @param engine - Engine to which the font belongs
-   * @param name - The name of font
+   * @param name - The name of font want to create
+   * @param fontUrl - The font url to register, if not, will use system font
    * @returns The font object has been create
    */
-  static createFromOS(engine: Engine, name: string = ""): Font {
+  static async create(engine: Engine, name: string, fontUrl: string = ""): Promise<Font> {
```
Comment (GuoLei1990) [inline packages/core/src/2d/text/Font.ts:20]: OS font shouldn't async.

The API of Font resource create:
- const systomFont = Font.createFromOS();
- const ttfFont = Resource.load("xxx/yyy/font.ttf");
- const customFont = Resource.load("xxx/yyy/font.font");

**Pull request merged**